File: web/src/message_find.test.ts
import { it, expect, beforeAll, vi } from "vitest";
import { readFileSync } from "node:fs";
import { initSync, find_text } from "./wasm/shep_mail_content";
import cases from "../../shared/find-cases.json";
import { MessageFind, type SearchHit } from "./message_find";
import { BrowserSettings } from "./model";
beforeAll(() =>
  initSync({
    module: new WebAssembly.Module(
      readFileSync(
        new URL("./wasm/shep_mail_content_bg.wasm", import.meta.url),
      ),
    ),
  }),
);
it("WASM preserves the desktop whitespace, Unicode, literal and UTF-16 contract", () => {
  for (const c of cases)
    expect(
      JSON.parse(find_text(JSON.stringify(c.blocks), c.query, c.match_case)),
    ).toEqual(c.hits);
});
it("coalesces pending searches and rejects old message, scope, query, error and close results", async () => {
  vi.useFakeTimers();
  const jobs: {
    blocks: string[];
    query: string;
    resolve: (hits: SearchHit[]) => void;
    reject: (error: Error) => void;
  }[] = [];
  const find = new MessageFind(
    (blocks, query) =>
      new Promise((resolve, reject) =>
        jobs.push({ blocks, query, resolve, reject }),
      ),
  );
  try {
    find.setSource("first", ["old"]);
    find.show();
    find.setQuery("old");
    await vi.advanceTimersByTimeAsync(100);
    expect(jobs).toHaveLength(1);
    find.setSource("second", ["latest"]);
    find.setQuery("discarded");
    find.setQuery("latest");
    await vi.advanceTimersByTimeAsync(100);
    expect(jobs).toHaveLength(1);
    jobs[0].reject(new Error("obsolete"));
    await vi.advanceTimersByTimeAsync(0);
    expect(find.error).toBeNull();
    expect(jobs).toHaveLength(2);
    expect(jobs[1].blocks).toEqual(["latest"]);
    expect(jobs[1].query).toBe("latest");
    jobs[1].resolve([
      { block: 0, start: 0, end: 2 },
      { block: 0, start: 2, end: 4 },
    ]);
    await vi.advanceTimersByTimeAsync(0);
    expect(find.status).toBe("1 of 2");
    find.next(true);
    expect(find.status).toBe("2 of 2");
    find.next();
    expect(find.status).toBe("1 of 2");
    find.toggleCase();
    await vi.advanceTimersByTimeAsync(100);
    find.close();
    jobs[2].resolve([{ block: 0, start: 0, end: 6 }]);
    await vi.advanceTimersByTimeAsync(0);
    expect(find.hits).toEqual([]);
    expect(find.open).toBe(false);
    find.show();
    await vi.advanceTimersByTimeAsync(100);
    jobs[3].reject(new Error("current"));
    await vi.advanceTimersByTimeAsync(0);
    expect(find.error).toContain("Retry Find");
    find.retry();
    await vi.advanceTimersByTimeAsync(100);
    jobs[4].resolve([]);
    await vi.advanceTimersByTimeAsync(0);
    expect(find.error).toBeNull();
    expect(find.status).toBe("No matches");
  } finally {
    find.dispose();
    vi.useRealTimers();
  }
});

it("new Find defaults preserve existing shortcut assignments and deliberate disabling", () => {
  try {
    vi.stubGlobal("localStorage", {
      getItem: () =>
        JSON.stringify({ version: 1, shortcuts: { move: "Control+f" } }),
    });
    expect(new BrowserSettings().read().shortcuts).toMatchObject({
      move: "Control+f",
      find: "",
    });
    vi.stubGlobal("localStorage", {
      getItem: () => JSON.stringify({ version: 1, shortcuts: { find: "" } }),
    });
    expect(new BrowserSettings().read().shortcuts.find).toBe("");
  } finally {
    vi.unstubAllGlobals();
  }
});

File: flutter/test/message_find_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:shep_mobile/data/message_search.dart';
import 'package:shep_mobile/model/message_find.dart';

void main() {
  test(
    'preview matches the shared desktop literal Unicode and wrapped whitespace contract',
    () async {
      final cases =
          jsonDecode(await File('../shared/find-cases.json').readAsString())
              as List;
      for (final c in cases) {
        expect(
          (await previewFind(
            List<String>.from(c['blocks']),
            c['query'],
            c['match_case'],
          )).map((h) => h.toJson()).toList(),
          c['hits'],
        );
      }
    },
  );
  test(
    'pending work coalesces and stale source/query/errors cannot restore a closed search',
    () async {
      final jobs = <(List<String>, String, Completer<List<SearchHit>>)>[];
      final find = MessageFind((blocks, query, _) {
        final c = Completer<List<SearchHit>>();
        jobs.add((blocks, query, c));
        return c.future;
      }, debounce: Duration.zero);
      addTearDown(find.dispose);
      Future<void> pump() => Future<void>.delayed(Duration.zero);
      find.setSource('first', ['old']);
      find.show();
      find.setQuery('old');
      await pump();
      expect(jobs, hasLength(1));
      find.setSource('second', ['latest']);
      find.setQuery('discarded');
      find.setQuery('latest');
      await pump();
      expect(jobs, hasLength(1));
      jobs[0].$3.completeError(StateError('obsolete'));
      await pump();
      expect(find.error, isNull);
      expect(jobs, hasLength(2));
      expect(jobs[1].$1, ['latest']);
      expect(jobs[1].$2, 'latest');
      jobs[1].$3.complete([const SearchHit(0, 0, 2), const SearchHit(0, 2, 4)]);
      await pump();
      expect(find.status, '1 of 2');
      find.next(true);
      expect(find.status, '2 of 2');
      find.next();
      expect(find.status, '1 of 2');
      find.toggleCase();
      await pump();
      find.close();
      jobs[2].$3.complete([const SearchHit(0, 0, 6)]);
      await pump();
      expect(find.hits, isEmpty);
      expect(find.open, false);
      find.show();
      await pump();
      jobs[3].$3.completeError(StateError('current'));
      await pump();
      expect(find.error, contains('Retry Find'));
      find.retry();
      await pump();
      jobs[4].$3.complete([]);
      await pump();
      expect(find.error, isNull);
      expect(find.status, 'No matches');
    },
  );
}
